OpenTabletDriver.Plugin/Output/Async/AsyncFilterHandler.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            this.scheduler = scheduler;
+            this.scheduler.Elapsed += SchedulerFunc;
+        }
+
+        private object asyncLock = new object();
+        private ITimer scheduler;
+        private HPETDeltaStopwatch asyncWatch = new HPETDeltaStopwatch(true);
+        private Action<ITabletReport> output;
+        private double reportMsAvg = 10.0;
+        private float frequency = 1000;
+        private bool penInRange, isDisposed;
+
+        public bool Enabled { get; set; }
+        public bool PenInRange
+        {
+            set
+            {
+                if (penInRange != value)
+                {
+                    penInRange = value;
+
+                    if (penInRange)
+                        scheduler.Start();
+                    else
+                        scheduler.Stop();
+                }
+            }
+            get => penInRange;
+        }
+
+        public IAsyncFilter AsyncFilter { get; set; }
+        public IFilter[] PreAsyncFilters { get; set; }
+        public float Frequency
+        {
+            set
+            {
+                frequency = value;
+                this.scheduler.Interval = 1000.0f / frequency;
+            }
+            get => frequency;
+        }
+
+        public void Bind(Action<ITabletReport> function)
+        {
+            output = function;
+        }
+
+        public void Process(ITabletReport report)
+        {
+            lock (asyncLock)
+            {
+                var reportSpan = asyncWatch.Restart();
+                reportMsAvg += (reportSpan.TotalMilliseconds - reportMsAvg) / 10.0;
+                PenInRange = true;
+
+                if (Enabled)
+                {
+                    var synthesizedReport = new SyntheticTabletReport(report);
+                    synthesizedReport.Position = ApplyPreAsyncFilters(synthesizedReport.Position);
+                    AsyncFilter.UpdateState(synthesizedReport, reportSpan);
+                }
+            }
+        }
+
+        public Vector2 ApplyPreAsyncFilters(Vector2 pos)
+        {
+            foreach (var filter in this.PreAsyncFilters)
+                pos = filter.Filter(pos);
+            return pos;
+        }
+
+        private void SchedulerFunc()
+        {
+            lock (asyncLock)
+            {
+                var reportSpan = asyncWatch.Elapsed;
+                if (WithinDelayTolerance(reportSpan.TotalMilliseconds, reportMsAvg) && PenInRange)
+                {
+                    var report = AsyncFilter.Filter(reportSpan);
+                    output(report);
+                }
+                else
+                {
+                    PenInRange = false;
+                }
+            }
+        }
+
+        public static bool WithinDelayTolerance(double reportMs, double reportMsAvg)
+        {
+            if (reportMsAvg < 3)
+                return reportMs < 3;
+            else
+                return reportMs < (reportMsAvg * 1.5);
```

Comment:
This function looks kinda sus
![image](https://user-images.githubusercontent.com/38468635/109690109-0a6b1480-7b97-11eb-9659-7e87eba4327c.png)
Maybe we can just simplify it to
```suggestion
            return (reportMs < 4.5f) | (reportMs < (reportMsAvg * 1.5));
```
?

Reply:
Is something like this fine?
```cs
            return reportMs < Math.Max(3, reportMsAvg * 1.5);
```

Reply:
Hmm, yeah, that looks even better